modules/ding_campaign/plugins/content_types/grid_campaign.inc

```diff
@@ -0,0 +1,143 @@
+<?php
+
+/**
+ * @file
+ * Ding Campaign CTools plugin file.
+ */
+
+$plugin = array(
+  'title' => t('Grid of campaigns'),
+  'description' => t('Display grid of campaigns'),
+  'single' => TRUE,
+  'defaults' => array(
+    'ding_campaign_count' => 6,
+    'ding_campaign_offset' => 0,
+  ),
+  'required context' => array(
+    new ctools_context_optional(t('Page node'), 'node'),
+    new ctools_context_optional(t('Related library'), 'node'),
+    new ctools_context_optional(t('Search term'), 'string'),
+    new ctools_context_optional(t('Taxonomy term'), array('entity:taxonomy_term', 'taxonomy_term')),
+  ),
+  'category' => t('Ding!'),
+);
+
+/**
+ * Implements hook_content_type_render().
+ */
+function ding_campaign_grid_campaign_content_type_render($subtype, $conf, $panel_args, $context) {
+  $block = new stdClass();
+  $context_data = array(
+    'page' => isset($context[0]->data) ? $context[0]->data : NULL,
+    'library' => isset($context[1]->data) ? $context[1]->data : NULL,
+    'search_term' => isset($context[2]->data) ? $context[2]->data : NULL,
+    'taxonomy_term' => isset($context[3]->data) ? $context[3]->data : NULL,
+  );
+  $count = $conf['ding_campaign_count'];
+  $offset = $conf['ding_campaign_offset'];
+  $more_link = $conf['ding_campaign_more_link'];
+  // Set default value if it was not set in panel settings.
+  $style = (!empty($conf['ding_campaign_style'])) ? $conf['ding_campaign_style'] : 'ding_campaign_full_width';
+
+  $block->content = ding_campaign_display($context_data, $count, $offset, $style, $subtype, $more_link);
+
+  return $block;
+}
+
+/**
+ * Implements hook_content_type_edit_form().
+ */
+function ding_campaign_grid_campaign_content_type_edit_form($form, &$form_state) {
+
+  $form['ding_campaign_count'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Campaign count'),
+    '#description' => t('The maximum number of campaigns to display'),
+    '#default_value' => (isset($form_state['conf']['ding_campaign_count'])) ? $form_state['conf']['ding_campaign_count'] : 3,
+    '#size' => 2,
+    '#maxlength' => 2,
+  );
+
+  $form['ding_campaign_offset'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Campaign offset'),
+    '#description' => t('Which campaign should we count from when returning campaigns. This is used to avoid displaying the same campaigns when using multiple campaign panes on the same panel. Use 0 to return the most relevant campaigns.'),
+    '#default_value' => (isset($form_state['conf']['ding_campaign_offset'])) ? $form_state['conf']['ding_campaign_offset'] : 0,
+    '#size' => 3,
+    '#maxlength' => 3,
```

Comment:
For validation consider [element_validate_integer_positive](https://api.drupal.org/api/drupal/includes%21form.inc/function/element_validate_integer_positive/7.x)